documentation/design/proposals/versioning_upgrade_refactoring.adoc

```diff
@@ -0,0 +1,74 @@
+    if we ever need to handle more sophisticated upgrades.
+ ** The addressspace's OAuth configuration becomes stale if the consoleservice is updated after the addressspace is created.   
+* No easy way for the user to determine major/minor/micro of the product being run.
+ ** The console displays a version - this is always an upstream version even in the downstream product.
+ ** No mechanism to determine the version from a command line/API.
+ ** (Downstream) No easy way to ensure that you are running a set of images from the same major/minor/micro release.
+* Many configuration items exposed as environment variables in api-server/address-space-controller/enmasse-controller     
+  deployments.  For Ansible/Bundle users configuration of these items is possible by a user (documentation is non existent).
+  For OLM, the situation is worse as the environment variables are baked into the OLM manifest.
+* Having three distinct install paths Ansible/Bundle/OLM increases the difficulty when reasoning about the defects in the
+  product.
+* Uninstall for OLM is broken (https://github.com/EnMasseProject/enmasse/issues/2863).
+
+== what
+
+=== Operator Responsible For Install
+
+The enmasse-operator will be given the responsibilities to deploy the api-server, address-space-controller, and service-broker and associated services, routes, roles, rolebindings etc. 
+
+It will also be responsible for reconciling the current state of those object against the desired state.
+
+Ansible and Bundle install methods will be maintained, but they will be much simplified.  They will merely need to deploy the operator itself (and any necessary role/rolebindings), the CRDs.  For backward compatibility, it could create an instance of the EnMasse CR.  For OLM, the metadata shipped will be much simplified.
+
+=== Operator Responsible For Upgrade
+
+The operator will have access to the CR's current version and actual version.  This information can be used to trigger
+upgrade specific processing.  After the upgrade is complete, the operator must change the spec.version to reflect the current running version.
+
+=== EnMasse CR
+
+There will be an EnMasse CRD with an instance of this CR representing EnMasse itself.
+
+The enmasse-operator will watch its own namespace for the creation of  the enmasse CR.  On its appearence, it will take care
+of the deployment.  Conversely, if the CR is removed it will be removed.  This will go towards fixing #2863.
+
+The CR's spec will carry the configuration items that are currently expressed as environment variables on the enmasse-operator, api-server, address-space-controller.   If the CR does not specify a configuration item, its default will be taken
+from the environment variables passed to the enmasse-operator itself.
+
+The CR will also carry the major/minor/micro version number of the version of EnMasse that created the record.  The status section will expose the current version.
```

Comment:
I think with OLM manifest gaining the ability to configure deployments, and migrating stuff to the enmasse-operator, the value of a toplevel resource like this is minimal IMO. It could function as a proxy for creating default instances of ConsoleService,AuthenticationService and IoTConfig for instance, they can also be created manually, so I'm not sure....

Reply:
The question is, how easy is this going to be to configure... are the permissions needed to install the operator going to be significantly higher than the rights needed to create/modify a CR with the configuration

Reply:
So assuming api-server, address-space-controller is gone, then what is it that really needs to be configured? Only the parameters of the enmasse-operator process itself, which you can't configure via a CR, because the operator must already be deployed in order for the CR to be processed.